---
title: "Final Report"
author: "Valeriia Ladyhina & Semyon Kupriyanov"
date: "`r Sys.Date()`"
output:  
  prettydoc::html_pretty:
    theme: architect
    highlight: lumen
---
# Introduction
*This R Markdown is part of bigger pipeline on analysis of RNA seq data obtained in experiment on effects of different concentrations of antibiotic novobiocin on gene expression in __Salmonella enterica subsp. enterica serovar Typhimurium str. 14028S__. Analysis was performed on publicly available transciptomics data that can be found __[here]('https://www.sciencedirect.com/science/article/pii/S2352340920301918#sec1')__.*

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Installation of required packages
*This part of pipeline includes few different packages, that have been installed on your computer, namely: __DESeq2, dplyr, ggplot2, gplots,clusterProfiler, WGCNA, knitr, KEGGREST,stringr__.*
```{r echo=FALSE, warning=FALSE, message=FALSE}
requiredPackages = c('DESeq2', 'dplyr', 'ggplot2','gplots', 'clusterProfiler','WGCNA','knitr','KEGGREST','stringr')
for(p in requiredPackages){
  if(!require(p,character.only = TRUE)) install.packages(p, repos = "http://cran.us.r-project.org" )
  library(p,character.only = TRUE)}
```
# DESeq2 analysis
*For this analysis we used two files: __transcript_count_matrix.csv__, obtained during previous steps of snalemake pipeline based on StringTie and __pheno.csv__ - file that describes our experimental data.*
```{r echo=FALSE, results='hide', message=FALSE, results='hide'}
countData <- read.csv("./transcript_count_matrix.csv", row.names="transcript_id")
colData <- read.csv('./data/pheno.csv', sep=";", row.names=1)
# Transform to factors
colData$Time <- factor(colData$Time)
colData$Treated <- factor(colData$Treated)
str(colData)
all(rownames(colData) %in% colnames(countData))
countData <- countData[, rownames(colData)]
all(rownames(colData) == colnames(countData))

# Create a DESeqDataSet from count matrix and labels

dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData, design = ~ Treated+Time+Treated:Time)

# Run the default analysis for DESeq2 and generate results table
dds <- DESeq(dds)
res <- results(dds)
# Sort by adjusted p-value and display
(resOrdered <- res[order(res$padj), ])
# Sort the results data frame by the padj and foldChange columns.
sorted = res[with(res, order(padj, -log2FoldChange)), ]
sorted.df = data.frame("id"=rownames(sorted),sorted)
```

*Next step is normalization of counts and saving of the result into file __norm-matrix-deseq2.txt__*

```{r echo=FALSE, results='hide'}
nc = counts(dds,normalized=TRUE)
dt = data.frame("id"=rownames(nc),nc)
```

*PCA analysis*

```{r echo=FALSE, results='hide'}
# The function rlogTransform returns a SummarizedExperiment object which contains the rlog-transformed
# values in its assay slot:

rld <- rlog(dds)
head(assay(rld))

plotPCA(rld, intgroup = c("Treated", "Time"))
```

*At this step we are searching for significant genes*

```{r echo=FALSE, results='hide'}
# Find significant genes
res_1 = as.data.frame(res)
res_1 = mutate(res_1, sig=ifelse(res_1$padj<0.05, "FDR<0.05", "Not_sig"))
res_1[which(abs(res_1$log2FoldChange)<1.5), "sig"] = "Not_sig"
res_2 = res_1[which(res_1$sig == "FDR<0.05"),] # 789 genes
sum(res_2$log2FoldChange>0) # 370 positive genes
sum(res_2$log2FoldChange<0) # 419 negative genes
```
*Volcano plot*

```{r echo=FALSE}
# Volcano_plot

ggplot(res_1, aes(log2FoldChange, -log10(padj)))+
  geom_point(aes(col=sig))+
  scale_color_manual(values = c("red", "black"))
```
```{r echo=FALSE, eval=FALSE}
# Check some parametrs
# A so-called MA plot provides a useful overview for an experiment with a two-group comparison:
plotMA(res, ylim = c(-8, 8))
```
```{r echo=FALSE,, eval=FALSE}
# The function plotDispEsts visualizes DESeq2вЂ™s dispersion estimates:
plotDispEsts(dds, ylim = c(1e-6, 1e1))
```
```{r echo=FALSE, eval=FALSE}
# Another useful diagnostic plot is the histogram of the p values
hist(res$pvalue, breaks=20, col="grey")
```

*DEG heatmap*

```{r echo=FALSE}
# Heatmap for DEG
matrix_deg = assay(rld)[rownames(res_2), ]
rownames(matrix_deg) = sapply(rownames(matrix_deg ), 
                    function(x) strsplit(x, split = "_")[[1]][2])
heatmap.2(matrix_deg, scale = "row",
          trace="none", dendrogram = "both")
```

*Negative and positive DEG were saved in files __negative_deg.txt__ and __positive_deg.txt__ respectively.*

# KEGG enrichment analysis

*Negative DEG*
```{r echo=FALSE, results='hold', message=FALSE, warning=FALSE}
# Download transformed deg-lists
deg_negative = read.table("New_rna_negative_deg.txt")
deg_positive = read.table("New_rna_positive_deg.txt")

# KEGG enrichmant analysis
kegg_deg_negative <- enrichKEGG(gene = deg_negative$V1,
                         organism  = 'seo',
                         pvalueCutoff = 0.05)

kegg_deg_positive <- enrichKEGG(gene = deg_positive$V1,
                         organism  = 'seo',
                         pvalueCutoff = 0.05)
# Search results
kable(head(kegg_deg_negative)) # Flagellar assembly # Oxidative phosphorylation
# Glycolysis / Gluconeogenesis # Bacterial chemotaxis
# Starch and sucrose metabolism

```

*Positive DEG*
```{r echo=FALSE}
kable(head(kegg_deg_positive)) 
```


```{r echo=FALSE}
# Download the date
df = read.table("norm-matrix-deseq2.txt", header = T, row.names = 1)
df = as.data.frame(t(df))
```
```{r echo=FALSE, results='hide'}
# Check the good genes
gsg_df = goodSamplesGenes(df, verbose = 3);
gsg_df$allOK
if (!gsg_df$allOK)
{
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg_df$goodGenes)>0)
    printFlush(paste("Removing genes:", paste(names(df)[!gsg_df$goodGenes], collapse = ", ")));
  if (sum(!gsg_df$goodSamples)>0)
    printFlush(paste("Removing samples:", paste(rownames(df)[!gsg_df$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  df = df[gsg_df$goodSamples, gsg_df$goodGenes]
}
```

# Sample clusterisation to detect outriers

```{r echo=FALSE}
# Make the samples clusterisation
sampleTree = hclust(dist(df), method = "average");
# Plot the sample tree: Open a graphic output window of size 12 by 9 inches
plot(sampleTree, main = "Sample clustering to detect outliers", sub="", xlab="", cex.lab = 1.5,
     cex.axis = 1.5, cex.main = 2)
```

*Added to clusters information regarding traits of data set*

```{r echo=FALSE}
# Download traits
traitData = read.csv('data/pheno.csv', sep=";", row.names=1)
# Samples clustering
sampleTree2 = hclust(dist(df), method = "average")
traitColors = numbers2colors(traitData, signed = F);
plotDendroAndColors(sampleTree, traitData,
                    groupLabels = names(traitData),
                    main = "Sample dendrogram and trait heatmap")
```
```{r echo=FALSE, results='hide', warning=FALSE}
# Pick soft threshold
powers = c(1:20)
cex1 = 0.8;
# Call the network topology analysis function
sft = pickSoftThreshold(df, powerVector = powers, verbose = 5)
```
*In this case we will use soft threashold in power 12*
```{r echo=FALSE}
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit,signed R^2",type="n",
     main = paste("Scale independence"));
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,cex=cex1,col="red");
# this line corresponds to using an R^2 cut-off of h
abline(h=0.80,col="red")
# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],
     xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",
     main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
```
```{r echo=FALSE, results='hide'}
# Build the NET
net = blockwiseModules(df, power = 12,
                       TOMType = "signed", minModuleSize = 30,
                       reassignThreshold = 0, mergeCutHeight = 0.25,
                       numericLabels = TRUE, pamRespectsDendro = FALSE,
                       saveTOMs = TRUE,
                       saveTOMFileBase = "typh_TOM",
                       verbose = 3)

table(net$colors) # 10 moduls + 1 zero
```

*Module clustering dendrogram* 

```{r echo=FALSE}
# Visualize
mergedColors = labels2colors(net$colors)
# Plot the dendrogram and the module colors underneath
plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]],
                    "Module colors",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)


# Some parametrs
moduleLabels = net$colors
moduleColors = labels2colors(net$colors)
MEs = net$MEs;
geneTree = net$dendrograms[[1]];
save(MEs, moduleLabels, moduleColors, geneTree,
     file = "wgcna.RData")
```
*Significance of modules and their correlation with traits using Spearman's correlation, we will count that resulta are significant if the are >0.7*

```{r echo=FALSE, warning=FALSE}
# Define modules significance and correlation with Traits

# Define numbers of genes and samples
nGenes = ncol(df);
nSamples = nrow(df);
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(df, moduleColors)$eigengenes
MEs = orderMEs(MEs0)
moduleTraitCor = cor(MEs, traitData, use = "p", method = "spearman");
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);

# Visual

# Will display correlations and their p-values
textMatrix = paste(signif(moduleTraitCor, 2), "\n(",
                   signif(moduleTraitPvalue, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)
par(mar = c(6, 8.5, 3, 3));
# Display the correlation values within a heatmap plot
labeledHeatmap(Matrix = moduleTraitCor,
               xLabels = names(traitData),
               yLabels = names(MEs),
               ySymbols = names(MEs),
               colorLabels = FALSE,
               colors = greenWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.text = 0.5,
               zlim = c(-1,1),
               main = paste("Module-trait relationships"))
```
*In our analysis we have five modules that are significant regarding treatment: brown, turquoise, blue, black, yellow; and two regarding time green and red*

*Plot that reveals relationships between eigengenes and traits*

```{r echo=FALSE}
# Define variable weight containing the weight column of datTrait
Treated = as.data.frame(traitData$Treated);
names(Treated) = "Treated"
# names (colors) of the modules
modNames = substring(names(MEs), 3)
geneModuleMembership = as.data.frame(cor(df, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");
geneTraitSignificance = as.data.frame(cor(df, Treated, use = "p", method = "spearman"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));
names(geneTraitSignificance) = paste("GS.", names(Treated), sep="");
names(GSPvalue) = paste("p.GS.", names(Treated), sep="");

# OK, Let's define significan modeles (cor>0.7)
# yellow, green, turquoise, black, brown, blue, red

colors_modules = c('yellow', 'green', 'turquoise', 
                   'black', 'brown', 'blue', 'red')
for (i in 1:length(colors_modules)){
  write.table(names(df)[moduleColors==colors_modules[i]], 
              file=paste0(colors_modules[i], "_result.txt"), 
              sep="\t", col.names=NA, quote=FALSE)
}

# Recalculate module eigengenes
MEs = moduleEigengenes(df, moduleColors)$eigengenes
# Isolate weight from the clinical traits
Treated = as.data.frame(traitData$Treated);
names(Treated) = "Treated"
# Add the weight to existing module eigengenes
MET = orderMEs(cbind(MEs, Treated))
# Plot the relationships among the eigengenes and the trait
par(cex = 0.9)
plotEigengeneNetworks(MET, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2), cex.lab = 0.8, xLabelsAngle
                      = 90)
```
*In this part we perform KEGG enrichment analysis of modules of interest*

```{r echo=FALSE, message=FALSE, warning=FALSE}
#####Let's make amazing KEGG analysis
yellow = read.table("New_rna_yellow_result.txt")
red = read.table("New_rna_red_result.txt")
black = read.table("New_rna_black_result.txt")
green = read.table("New_rna_green_result.txt")
turquoise = read.table("New_rna_turquoise_result.txt")
brown = read.table("New_rna_brown_result.txt")
blue = read.table("New_rna_blue_result.txt")


kegg_yellow <- enrichKEGG(gene = yellow$V1,
                          organism  = 'seo',
                          pvalueCutoff = 0.05)

kegg_red <- enrichKEGG(gene = red$V1,
                       organism  = 'seo',
                       pvalueCutoff = 0.05)

kegg_black <- enrichKEGG(gene = black$V1,
                         organism  = 'seo',
                         pvalueCutoff = 0.05)

kegg_green <- enrichKEGG(gene = green$V1,
                         organism  = 'seo',
                         pvalueCutoff = 0.05)

kegg_turquoise <- enrichKEGG(gene = turquoise$V1,
                             organism  = 'seo',
                             pvalueCutoff = 0.05)

kegg_blue <- enrichKEGG(gene = blue$V1,
                        organism  = 'seo',
                        pvalueCutoff = 0.05)

kegg_brown <- enrichKEGG(gene = brown$V1,
                         organism  = 'seo',
                         pvalueCutoff = 0.05)
```

*First let's take a look on modules that were sighnificant in relation towards novobiocin treatment*

*Module 1*

```{r echo=FALSE}
kable(head(kegg_turquoise)[2:6])

```

*Module 2*

```{r echo=FALSE}
kable(head(kegg_black)) 
```

*Module 3*

```{r echo=FALSE}
kable(head(kegg_yellow))
```
 
*Module 4*
   
```{r echo=FALSE}
kable(head(kegg_brown))
```
*Module 5*

```{r echo=FALSE}
kable(head(kegg_blue))
```
*And now we will take a look on modules related to time*

*Module 1*

```{r echo=FALSE}
kable(head(kegg_red)[2:6])
```

*Module 2*

```{r echo=FALSE}
kable(head(kegg_green)) 
```

*Now we will perform visualisation of involed KEGG paths and you will find results in your browser*

```{r echo=FALSE}
# Visualization of paths that have been found in both DEGs and modules associated with Treated 
browseKEGG(kegg_black, 'seo02020')
#
browseKEGG(kegg_brown, 'seo00660')
browseKEGG(kegg_brown, 'seo00053')
browseKEGG(kegg_brown, 'seo01210')
#
browseKEGG(kegg_turquoise, 'seo00010')
browseKEGG(kegg_turquoise, 'seo00520')
```
```{r echo=FALSE,message=FALSE,warning=FALSE, results='hide'}

# Download the date
df = read.table("norm-matrix-deseq2.txt", header = T, row.names = 1)
df = as.data.frame(t(df))

# Check the good genes
gsg_df = goodSamplesGenes(df, verbose = 3);
gsg_df$allOK
if (!gsg_df$allOK)

  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg_df$goodGenes)>0)
    printFlush(paste("Removing genes:", paste(names(df)[!gsg_df$goodGenes], collapse = ", ")));
  if (sum(!gsg_df$goodSamples)>0)
    printFlush(paste("Removing samples:", paste(rownames(df)[!gsg_df$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  df = df[gsg_df$goodSamples, gsg_df$goodGenes]

# Make the samples clusterisation
sampleTree = hclust(dist(df), method = "average");

traitData = read.csv('data/pheno.csv', sep=";", row.names=1)
# Samples clustering
sampleTree2 = hclust(dist(df), method = "average")
traitColors = numbers2colors(traitData, signed = F);

powers = c(1:20)
cex1 = 0.8;
# Call the network topology analysis function
sft = pickSoftThreshold(df, powerVector = powers, verbose = 5)

# Build the NET
net = blockwiseModules(df, power = 12,
                       TOMType = "signed", minModuleSize = 30,
                       reassignThreshold = 0, mergeCutHeight = 0.25,
                       numericLabels = TRUE, pamRespectsDendro = FALSE,
                       saveTOMs = TRUE,
                       saveTOMFileBase = "typh_TOM",
                       verbose = 3)

table(net$colors) # 10 moduls + 1 zero

# Visualize
mergedColors = labels2colors(net$colors)


# Some parametrs
moduleLabels = net$colors
moduleColors = labels2colors(net$colors)
MEs = net$MEs;
geneTree = net$dendrograms[[1]];
save(MEs, moduleLabels, moduleColors, geneTree,
     file = "wgcna.RData")

nGenes = ncol(df);
nSamples = nrow(df);
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(df, moduleColors)$eigengenes
MEs = orderMEs(MEs0)
moduleTraitCor = cor(MEs, traitData, use = "p", method = "spearman");
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples);

# Visual

# Will display correlations and their p-values
textMatrix = paste(signif(moduleTraitCor, 2), "\n(",
                   signif(moduleTraitPvalue, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)
par(mar = c(6, 8.5, 3, 3));

# Define variable weight containing the weight column of datTrait
Treated = as.data.frame(traitData$Treated);
names(Treated) = "Treated"
# names (colors) of the modules
modNames = substring(names(MEs), 3)
geneModuleMembership = as.data.frame(cor(df, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
names(geneModuleMembership) = paste("MM", modNames, sep="");
names(MMPvalue) = paste("p.MM", modNames, sep="");
geneTraitSignificance = as.data.frame(cor(df, Treated, use = "p", method = "spearman"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));
names(geneTraitSignificance) = paste("GS.", names(Treated), sep="");
names(GSPvalue) = paste("p.GS.", names(Treated), sep="");

# OK, Let's define significan modeles (cor>0.7)
# yellow, green, turquoise, black, brown, blue, red

colors_modules = c('yellow', 'green', 'turquoise', 
                   'black', 'brown', 'blue', 'red')
for (i in 1:length(colors_modules)){
  write.table(names(df)[moduleColors==colors_modules[i]], 
              file=paste0(colors_modules[i], "_result.txt"), 
              sep="\t", col.names=NA, quote=FALSE)
}

# Recalculate module eigengenes
MEs = moduleEigengenes(df, moduleColors)$eigengenes
# Isolate weight from the clinical traits
Treated = as.data.frame(traitData$Treated);
names(Treated) = "Treated"
# Add the weight to existing module eigengenes
MET = orderMEs(cbind(MEs, Treated))
```

```{r echo=FALSE}
write.table(rownames(geneModuleMembership[which(geneModuleMembership$MMblack > 0.9),]), 
            file="black_hub.txt", 
            sep="\t", col.names=NA, quote=FALSE)

write.table(rownames(geneModuleMembership[which(geneModuleMembership$MMbrown > 0.9),]), 
            file="brown_hub.txt", 
            sep="\t", col.names=NA, quote=FALSE)

write.table(rownames(geneModuleMembership[which(geneModuleMembership$MMyellow > 0.9),]), 
            file="yellow_hub.txt", 
            sep="\t", col.names=NA, quote=FALSE)

write.table(rownames(geneModuleMembership[which(geneModuleMembership$MMturquoise > 0.9),]), 
            file="turquoise_hub.txt", 
            sep="\t", col.names=NA, quote=FALSE)


# Time hubs
write.table(rownames(geneModuleMembership[which(geneModuleMembership$MMgreen > 0.9),]), 
            file="green_hub.txt", 
            sep="\t", col.names=NA, quote=FALSE)

write.table(rownames(geneModuleMembership[which(geneModuleMembership$MMred > 0.9),]), 
            file="red_hub.txt", 
            sep="\t", col.names=NA, quote=FALSE)

write.table(rownames(geneModuleMembership[which(geneModuleMembership$MMblue > 0.9),]), 
            file="blue_hub.txt", 
            sep="\t", col.names=NA, quote=FALSE)
```
```{r echo=FALSE}
# Download huns with transformed names
yellow_hub = read.table("New_rna_yellow_hub.txt")
red_hub = read.table("New_rna_red_hub.txt")
black_hub = read.table("New_rna_black_hub.txt")
green_hub = read.table("New_rna_green_hub.txt")
turquoise_hub = read.table("New_rna_turquoise_hub.txt")
brown_hub = read.table("New_rna_brown_hub.txt")
blue_hub = read.table("New_rna_blue_hub.txt")
```

*KEGG analysis of treated hubs*
```{r echo=FALSE, message=FALSE, warning=FALSE}
# KEGG analysis of Treated hubs
kegg_brown_hub = enrichKEGG(gene = brown_hub$V1,
                            organism  = 'seo',
                            pvalueCutoff = 0.05)
kable(head(kegg_brown_hub))
# C5-Branched dibasic acid metabolism
# Valine, leucine and isoleucine biosynthesis
# 2-Oxocarboxylic acid metabolism
# Biosynthesis of amino acids

kegg_black_hub = enrichKEGG(gene = black_hub$V1,
                            organism  = 'seo',
                            pvalueCutoff = 0.05)
kable(head(kegg_black_hub))
# Two-component system

kegg_yellow_hub = enrichKEGG(gene = yellow_hub$V1,
                             organism  = 'seo',
                             pvalueCutoff = 0.05)
kable(head(kegg_yellow_hub))
#Lipopolysaccharide biosynthesis

kegg_turquoise_hub = enrichKEGG(gene = turquoise_hub$V1,
                                organism  = 'seo',
                                pvalueCutoff = 0.05)
kable(head(kegg_turquoise_hub))
# Pyruvate metabolism
#  Microbial metabolism in diverse environments
# Citrate cycle (TCA cycle)
#  Glycolysis / Gluconeogenesis
# ABC transporters
# Carbon metabolism
```

*Now we will save pathways of interest based on WGCNA analysis related to treatment with novobiocin*

```{r echo=FALSE}
# Next we choose pathways from DESeq2 and WGCNA Treated modules

# Time course vizual
kegg_black_hub[1]$geneID # Two-component system

kegg_brown_hub[1]$geneID # C5-Branched dibasic acid metabolism

kegg_brown_hub[3]$geneID # 2-Oxocarboxylic acid metabolism

kegg_turquoise_hub[4]$geneID # Glycolysis / Gluconeogenesis

two_component_system = strsplit(kegg_black_hub[1]$geneID, "/")[[1]]
c5_branched_dibasic_acid_metabolism = strsplit(kegg_brown_hub[1]$geneID, "/")[[1]]
oxocarboxylic_acid_metabolism = strsplit(kegg_brown_hub[3]$geneID, "/")[[1]]
glycolysis_gluconeogenesis = strsplit(kegg_turquoise_hub[4]$geneID, "/")[[1]]

write.table(two_component_system, 
            file="two_component_system.txt", 
            sep="\t", col.names=NA, quote=FALSE)

write.table(c5_branched_dibasic_acid_metabolism, 
            file="c5_branched_dibasic_acid_metabolism.txt", 
            sep="\t", col.names=NA, quote=FALSE)

write.table(oxocarboxylic_acid_metabolism, 
            file="oxocarboxylic_acid_metabolism.txt", 
            sep="\t", col.names=NA, quote=FALSE)

write.table(glycolysis_gluconeogenesis, 
            file="glycolysis_gluconeogenesis.txt", 
            sep="\t", col.names=NA, quote=FALSE)
```

*Visualisation of modules eigenes*

```{r echo=FALSE}

# MODULES EIGENES VISUALISATION
# Treated modules
me_tr = MEs[c('MEblack', 'MEbrown', 'MEyellow', 'MEturquoise')]
me_tr$Time = c(10, 10, 10,
                   20, 20, 20,
                   60, 60, 60,
                   10, 10, 10,
                   20, 20, 20,
                   60, 60, 60)
me_tr$Treated = c(1, 1, 1,
                      1, 1, 1,
                      1, 1, 1,
                      0, 0, 0,
                      0, 0, 0,
                      0, 0, 0)
me_tr$Treated = factor(me_tr$Treated)
me_tr = me_tr %>% 
  group_by(Treated, Time) %>% 
  summarise_all(funs(mean))


ggplot(me_tr, aes(x=Time, y=MEblack, col=Treated))+
  geom_line()+
  theme_light()
ggplot(me_tr, aes(x=Time, y=MEbrown, col=Treated))+
  geom_line()+
  theme_light()
ggplot(me_tr, aes(x=Time, y=MEyellow, col=Treated))+
  geom_line()+
  theme_light()
ggplot(me_tr, aes(x=Time, y=MEturquoise, col=Treated))+
  geom_line()+
  theme_light()

# Time modules
me_ti = MEs[c('MEgreen', 'MEred', 'MEblue')]
me_ti$Time = c(10, 10, 10,
               20, 20, 20,
               60, 60, 60,
               10, 10, 10,
               20, 20, 20,
               60, 60, 60)
me_ti$Treated = c(1, 1, 1,
                  1, 1, 1,
                  1, 1, 1,
                  0, 0, 0,
                  0, 0, 0,
                  0, 0, 0)
me_ti$Treated = factor(me_ti$Treated)
me_ti = me_ti %>% 
  group_by(Treated, Time) %>% 
  summarise_all(funs(mean))

ggplot(me_ti, aes(x=Time, y=MEred, col=Treated))+
  geom_line()+
  theme_light()
ggplot(me_ti, aes(x=Time, y=MEgreen, col=Treated))+
  geom_line()+
  theme_light()
ggplot(me_ti, aes(x=Time, y=MEblue, col=Treated))+
  geom_line()+
  theme_light()
```

*Visualisation of pathways'genes by their PC1*

```{r echo=FALSE}
# We will visualise pathway's genes by their PC1

# visual function
get_dynamic_graph = function(path_to, c=-1, name="Pathway"){
  c5 = read.table(path_to)
  sub_c5 = df[,c5$V1]
  colnames(sub_c5) = sapply(colnames(sub_c5),
                            function(x) strsplit(x, "-")[[1]][2])
  sub_c5$Time = c(10, 10, 10,
                  20, 20, 20,
                  60, 60, 60,
                  10, 10, 10,
                  20, 20, 20,
                  60, 60, 60)
  
  sub_c5$Treated = c(1, 1, 1,
                     1, 1, 1,
                     1, 1, 1,
                     0, 0, 0,
                     0, 0, 0,
                     0, 0, 0)
  sub_c5$Treated = factor(sub_c5$Treated)
  # PCA
  pca_sub_c5 <- prcomp(scale(sub_c5[-c(ncol(sub_c5)-1,ncol(sub_c5))]))
  #summary(pca_sub_tcs)
  sub_c5$PC1 <- (data.frame(predict(pca_sub_c5, 
                                    newdata = sub_c5[-c(ncol(sub_c5)-1,
                                                        ncol(sub_c5))]))$PC1)*(c)
  
  sub_c5_tr = sub_c5 %>% 
    group_by(Treated, Time) %>% 
    summarise_all(funs(mean))
  
  ggplot(sub_c5_tr, aes(x=Time, y=PC1, col=Treated))+
    geom_line()+
    ggtitle(name)+
    theme_light()
}
```

```{r echo=FALSE}
# two_component_system.txt
get_dynamic_graph("New_two_component_system.txt",1, 
                  "Two component system")

# c5_branched_dibasic_acid_metabolism 
get_dynamic_graph("New_c5_branched_dibasic_acid_metabolism.txt", 1,
                  "C5-branched dibasic acid metabolism")

# oxocarboxylic_acid_metabolism  
get_dynamic_graph("New_oxocarboxylic_acid_metabolism.txt", 1,
                  "2-oxocarboxylic acid metabolism")

# glycolysis_gluconeogenesis
get_dynamic_graph("New_glycolysis_gluconeogenesis.txt", 1,
                  "Glycolysis/Gluconeogenesis")
```